automarker_2/automarker_app/lib/steps.py

```diff
@@ -626,9 +633,8 @@ def run(self, project_uri, clone_dir_path, self_test, config, fail_fast):
         if len(stderr):
             self.set_outcome(STEP_STATUS_ERROR, message=stderr)
         else:
-            command = (
-                f"{clone_dir_path/'automarker_venv'/'bin'/'pip'} install --upgrade pip"
-            )
+            python_executable_path_virtual_env = f'{get_python_executable_path()}/python' if is_current_os_windows() else f'{get_python_executable_path()}/pip'
```

Comment:
we can cache the value of `get_python_executable_path()` so that we can reuse it in the else block to keep our code tidy. Plus we're repeating calling "is_current_os_windows"

```python
# atop:
def get_python_executable_path():
    if platform.system() == "Windows":
        return 'Scripts/python.exe'
    return  'bin/python'

# inside PythonCreateVirtualEnv (here)
command = f"{clone_dir_path/ 'automarker_venv'/ get_python_executable_path()} -m pip install --upgrade pip"
```

We can use the same approach when we want to pip install, no reason why we should differentiate how we're calling it in `def run(**)`

```python
# like this
command = f"{clone_dir_path/ 'automarker_venv'/ get_python_executable_path()} -m pip install -r {clone_dir_path/'requirements.txt'}"
```